Use new tax classes for taxonomic summarization

bluegenes opened the pull request:
## Taxonomy Refactor Overview

In an attempt to allow usage of NCBI taxid (motivation: CAMI benchmarking) and alternate hierarchical taxonomic ranks (motivation: LINS), I ended up refactoring the taxonomy code in a four-PR series. Taxonomic summarization results should not change. Minor caveat: I was previously obtaining `query_bp` in a hacky manner to allow gather <4.4 results. The class methods are more robust, and I'd like to stop supporting gather <4.4 results. To allow this, I had to add the `query_bp`, `ksize`, and `scaled` columns into some testing results to keep tests functioning.

1. https://github.com/sourmash-bio/sourmash/pull/2437 modifies `LineagePair` from a two-item `collections.namedtuple` to a three-item `typing.NamedTuple` containing an additional field, `taxid`, for storing NCBI taxid information. It also introduces classes (`BaseLineageInfo`, `RankLineageInfo`), which move lineage manipulation (from `lca_utils.py`) to class methods in order to support robust summarization across compatible lineages (lineages of same hierarchical ranks). To ensure these can be used as dictionary keys, these classes are frozen.

2. https://github.com/sourmash-bio/sourmash/pull/2439 introduces classes that facilitate reading, summarization, and writing of gather results. First, it updates three prior `collections.namedtuple`s to `dataclasses`  used for storing information about the gather query (`QueryInfo`), summarized gather information for metagenome queries (`SummarizedGatherResult`) and classification information for genome queries (`ClassificationResult`). It introduces three new classes for reading and manipulating gather results. `GatherRow`, is used for reading a each row from a gather file and automatically checking for required columns. `TaxResult` is used for storing a single row from gather file, optionally (and ideally) with taxonomic information, stored as `LineageInfo` class from PR 1. `QueryTaxResult` is used for storing all `TaxResult`s associated with a single query. `QueryTaxResult` add methods to replicate the summarization previously done within `summarize_gather_at` in `tax_utils.py` and the classification thresholding in `genome` within `__main__.py`.

3. https://github.com/sourmash-bio/sourmash/pull/2443 replaces the actual taxonomic summarization code in `tax/__main__.py` with code that uses the new classes. Modifies gather loading code to read using `GatherRow`, `TaxResult`, and `QueryTaxResult`.

4. https://github.com/sourmash-bio/sourmash/pull/2446 removes old, unused functions that are rendered redundant by the new classes. Also removes associated tests.

## Additional details for this PR (#2443) 

Make use of `LineagePair`, `LineageInfo` classes (#2371) and `GatherRow`, `SummarizedGatherResult`, `ClassificationResult`, `TaxResult`, and `QueryTaxResult` classes (#2439) to do taxonomic summarization. 

Replaces old code where it's easy to see changes, leaves larger cleanup for #2446 where it might make this PR confusing.

Benefits of this refactor: safer/more robust taxonomic summarization 
  - innately checks that required gather columns are present when loading each gather csv
  - only allows summarization over gather results from same query, stores all info for a given query together.
  -  only allows summarization/comparison between compatible lineages (same hierarchical ranks)
  - more robust/simplified tracking of results with missed & skipped taxonomic identifiers + perfect matches
  - where multi-query outputs would be misleading, prevent users from producing these outputs when summarizing multiple queries
  - **enables future additions**:
    -  NCBI taxid can be included and reported during summarization (motivating reason: CAMI)
    -  Alternate rank systems should work with all the functions tested for named ranks (motivating reason: LINS)
 
Neutral changes:
- Some error messages changed slightly, because information was stored slightly differently/was differently available
- requires sourmash gather >4.4, which contains columns needed for robust query compatibility checks 
- ANI is estimated by default, since we now require all columns needed for ANI estimation
- added ksize, scaled columns to test csv files, where needed
- fixed minor error introduced in #2371, where an extra 0 was added in the `query_bp` column of `test1.gather.csv`

bluegenes committed:
init new LineagePair and LineageInfo classes

bluegenes committed:
test new LineagePair,BaseLineageInfo,RankLineageInfo

bluegenes committed:
fix

bluegenes committed:
add v4.4 columns that will be required

bluegenes committed:
Merge branch 'latest' into upd-lineage-utils

bluegenes committed:
make the newly added query_bp info in test1.gather.csv work with existing tests

bluegenes committed:
rename test1.gather_ani.csv to test1.gather_old.csv to reflect its new usage

bluegenes committed:
moar lineage tests

bluegenes committed:
test remaining codecov misses

bluegenes committed:
whoops, one last codecov miss

bluegenes committed:
add tax summarization classes; rename old NamedTuples to avoid breakage

bluegenes committed:
finish renaming

bluegenes committed:
add tests for new classes

bluegenes committed:
upd

bluegenes committed:
more tests; move status checking into ClassificationResult

bluegenes committed:
human summary dict tests

bluegenes committed:
add value checks and tests for SGR,CR classes

bluegenes committed:
test make_full_summary

bluegenes committed:
use f_unique, not f_weighted to preserve current functionality

bluegenes committed:
test no ranks

bluegenes committed:
test make_kreport_results

bluegenes committed:
krona fns that use the new dataclasses

bluegenes committed:
init changes to tax main

bluegenes committed:
upd load_gather fns

bluegenes committed:
init metagenome changes to use new framework

bluegenes committed:
Merge branch 'latest' into upd-tax-summarization

bluegenes committed:
Merge branch 'upd-tax-summarization' into use-new-tax-classes

bluegenes committed:
fix lineagepair

bluegenes committed:
Merge branch 'upd-tax-summarization' into use-new-tax-classes

bluegenes committed:
fix kreport num_bp; upd some err msg

bluegenes committed:
add ksize, scaled to mimic v4.4+ output

bluegenes committed:
upd krona writing

bluegenes committed:
use RankLineageInfo() instead of () for unclassified/empty tax

bluegenes committed:
display lineage within aggregate; all metagenome tests working

bluegenes committed:
init changes for genome, annotate

bluegenes committed:
fix annotate header; make sure load_gather doesnt ask for query_name before we know it exists

bluegenes committed:
enable lineage_csv output from genome

bluegenes committed:
now fail if gather <4.4

bluegenes committed:
bp is int, not float

bluegenes committed:
upd write classif

bluegenes committed:
add ksize and scaled to mimic v4.4

bluegenes committed:
use fraction, not f_weighted to match prior functionality

bluegenes committed:
test name_at_rank, as_lineage_dict

bluegenes committed:
test metagenome single-query output format restrictions

bluegenes committed:
test single query output formats; multi q krona

bluegenes committed:
cleaner

bluegenes commented:
ok @ctb, here's the PR where I actually use #2371 and #2439 -- ready for review. 

Tests passing, but I'm somehow missing codecov target - is this because of deletions/refactoring?

bluegenes committed:
Merge branch 'latest' into use-new-tax-classes

bluegenes committed:
Merge branch 'latest' into use-new-tax-classes

ctb commented on src/sourmash/tax/tax_utils.py:
```diff
@@ -953,17 +1146,19 @@ def write_lineage_sample_frac(sample_names, lineage_dict, out_fp, *, format_line
     for lin, sampleinfo in sorted(lineage_dict.items()):
-        if format_lineage:
-            lin = display_lineage(lin)
+        # if format_lineage:
```
suggest removing.

ctb replied:
(rather than uncommenting)

ctb commented on src/sourmash/tax/tax_utils.py:
```diff
@@ -622,7 +642,36 @@ def summarize_gather_at(rank, tax_assign, gather_results, *, skip_idents = [],
+        notify(f'of {total_taxresults} gather results, lineage assignments for {total_n_missed} results were missed.')
+        #notify(f'of {total_taxresults} gather results, missed lineage assignments for {total_n_missed)} results.')
+        notify(f'The following are missing from the taxonomy information: {", ".join(ident_missed)}')
+    # skipping not actually enabled? do we want to enable?
```
suggest removing skipping code completely.

ctb commented on src/sourmash/tax/tax_utils.py:
```diff
@@ -1703,25 +1903,25 @@ def __post_init__(self):
-        elif containment_threshold and self.fraction >= containment_threshold:
-        #elif containment_threshold and self.f_weighted_at_rank >= containment_threshold:
+        elif containment_threshold is not None and self.fraction >= containment_threshold:
+        #elif containment_threshold is not None and self.f_weighted_at_rank >= containment_threshold:
```
remove?

bluegenes replied:
modified - at some point, I'd like to chat about potentially using weighted results instead of unique results for ordering ..

ctb commented on tests/test_tax.py:
```diff
@@ -1901,45 +1997,28 @@ def test_genome_ani_threshold(runtmp):
     print(c.last_result.err)
-    assert "WARNING: classifying query test1 at desired rank species does not meet query ANI/AAI threshold 1.0" in c.last_result.err
-    assert "test1,below_threshold,species,0.089,d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella;s__Prevotella copri,md5,test1.sig,0.057,444000.0,0.9247805047263588" in c.last_result.out
+    # assert "WARNING: classifying query test1 at desired rank species does not meet query ANI/AAI threshold 1.0" in c.last_result.err
```
remove?

ctb commented on tests/test_tax.py:
```diff
@@ -1836,7 +1931,8 @@ def test_genome_over100percent_error(runtmp):
     assert runtmp.last_result.status == -1
-    assert "ERROR: The tax summary of query 'test1' is 1.1, which is > 100% of the query!!" in runtmp.last_result.err
+    assert "fraction is > 100% of the query! This should not be possible." in runtmp.last_result.err
+    # assert "ERROR: The tax summary of query 'test1' is 1.1, which is > 100% of the query!!" in runtmp.last_result.err
```
remove?

ctb commented on tests/test_tax.py:
```diff
@@ -1971,12 +2050,16 @@ def test_genome_ani_lemonade_classify(runtmp):
 
     output = c.last_result.out
-    assert 'MAG3_1             5.3%     91.0%  d__Bacteria;p__Bacteroidota;c__Chlorobia;o__Chlorobiales;f__Chlorobiaceae;g__Prosthecochloris;s__Prosthecochloris vibrioformis' in output
+    # assert 'MAG3_1            5.3%     91.0%  d__Bacteria;p__Bacteroidota;c__Chlorobia;o__Chlorobiales;f__Chlorobiaceae;g__Prosthecochloris;s__Prosthecochloris vibrioformis' in output
```
remove?

ctb reviewed (approved):
Nice work!

bluegenes committed:
cleanup taxonomy code after refactor (#2446)

## Taxonomy Refactor Overview

In an attempt to allow usage of NCBI taxid (motivation: CAMI
benchmarking) and alternate hierarchical taxonomic ranks (motivation:
LINS), I ended up refactoring the taxonomy code in a four-PR series.
Taxonomic summarization results should not change. Minor caveat: I was
previously obtaining `query_bp` in a hacky manner to allow gather <4.4
results. The class methods are more robust, and I'd like to stop
supporting gather <4.4 results. To allow this, I had to add the
`query_bp`, `ksize`, and `scaled` columns into some testing results to
keep tests functioning.

1. https://github.com/sourmash-bio/sourmash/pull/2437 modifies
`LineagePair` from a two-item `collections.namedtuple` to a three-item
`typing.NamedTuple` containing an additional field, `taxid`, for storing
NCBI taxid information. It also introduces classes (`BaseLineageInfo`,
`RankLineageInfo`), which move lineage manipulation (from
`lca_utils.py`) to class methods in order to support robust
summarization across compatible lineages (lineages of same hierarchical
ranks). To ensure these can be used as dictionary keys, these classes
are frozen.

2. https://github.com/sourmash-bio/sourmash/pull/2439 introduces classes
that facilitate reading, summarization, and writing of gather results.
First, it updates three prior `collections.namedtuple`s to `dataclasses`
used for storing information about the gather query (`QueryInfo`),
summarized gather information for metagenome queries
(`SummarizedGatherResult`) and classification information for genome
queries (`ClassificationResult`). It introduces three new classes for
reading and manipulating gather results. `GatherRow`, is used for
reading a each row from a gather file and automatically checking for
required columns. `TaxResult` is used for storing a single row from
gather file, optionally (and ideally) with taxonomic information, stored
as `LineageInfo` class from PR 1. `QueryTaxResult` is used for storing
all `TaxResult`s associated with a single query. `QueryTaxResult` add
methods to replicate the summarization previously done within
`summarize_gather_at` in `tax_utils.py` and the classification
thresholding in `genome` within `__main__.py`.

3. https://github.com/sourmash-bio/sourmash/pull/2443 replaces the
actual taxonomic summarization code in `tax/__main__.py` with code that
uses the new classes. Modifies gather loading code to read using
`GatherRow`, `TaxResult`, and `QueryTaxResult`.

4. https://github.com/sourmash-bio/sourmash/pull/2446 removes old,
unused functions that are rendered redundant by the new classes. Also
removes associated tests.

## Additional details for this PR (#2446) 

- Delete old functions that aren't used outside of taxonomic
summarization + associated tests
  - Including old `namedtuple`s: `QueryInf`, `SumGathInf`, `ClassInf`
- Make sure any old comments/documentation make it into new code
- Don't use unnecessary empty `()` for dataclasses

bluegenes committed:
Merge branch 'latest' into use-new-tax-classes

bluegenes committed:
apply suggestions from code review

bluegenes commented:
@ctb  Is this ok to merge with the codecov failure? I thought it would go away after merging #2446, but coverage still decreased (-0.11%). I assume this is a result of net deletions?

ctb commented:
there's no uncovered code, right?! So sure, it's fine!
